app/database/group_group_store.go

```diff
@@ -22,3 +27,53 @@ func (s *GroupGroupStore) WhereUserIsMember(user *User) *DB {
 func (s *GroupGroupStore) createNewAncestors() {
 	s.DataStore.createNewAncestors("groups", "Group")
 }
+
+var ErrRelationAlreadyExists = errors.New("the relation already exists")
+var ErrRelationCycle = errors.New("a group cannot become an ancestor of itself")
+
+const groupsRelationsLockTimeout = 3 * time.Second
+
+func (s *GroupGroupStore) CreateRelation(parentGroupID, childGroupID int64) (err error) {
+	s.mustBeInTransaction()
+	defer recoverPanics(&err)
+
+	mustNotBeError(s.WithNamedLock(s.tableName, groupsRelationsLockTimeout, func(store *DataStore) (err error) {
+		var rows []interface{}
+		mustNotBeError(s.WithWriteLock().
+			Select("ID").
+			Where("idGroupChild = ? AND idGroupParent = ?", childGroupID, parentGroupID).
+			Limit(1).
+			Scan(&rows).Error())
+		if len(rows) > 0 {
+			return ErrRelationAlreadyExists
+		}
+
+		mustNotBeError(store.GroupAncestors().
+			WithWriteLock().
+			Select("ID").
+			Where("idGroupChild = ? AND idGroupAncestor = ?", parentGroupID, childGroupID).
+			Limit(1).
+			Scan(&rows).Error())
+		if len(rows) > 0 {
+			return ErrRelationCycle
+		}
+
+		groupGroupStore := store.GroupGroups()
+
+		mustNotBeError(groupGroupStore.db.Exec(
+			"SET @maxIChildOrder = IFNULL((SELECT MAX(iChildOrder) FROM `groups_groups` WHERE `idGroupParent` = ? FOR UPDATE), 0)",
+			parentGroupID).Error)
+
+		mustNotBeError(groupGroupStore.retryOnDuplicatePrimaryKeyError(func(db *DB) error {
```

Comment:
Shouldn't `retryOnDuplicatePrimaryKeyError` take care of providing the ID directly as this is the only usage of this function?

Reply:
I have done it this way because (imho) DB should not depend on DataStore